src/ganted2radius.go

```diff
@@ -0,0 +1,246 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"io/fs"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/klauspost/compress/zstd"
+	"golang.org/x/net/context"
+	"layeh.com/radius"
+	"layeh.com/radius/rfc2865"
+	"layeh.com/radius/rfc2866"
+)
+
+func compressFile(filepath string) error {
+	// open the file
+	file, err := os.Open(filepath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	// create the compressed file, err if it already exists
+	compressedFilepath := filepath + ".zst"
+	compressedFile, err := os.Create(compressedFilepath)
+	if err != nil {
+		return err
+	}
+	defer compressedFile.Close()
+	// create the zstd writer
+	zw, err := zstd.NewWriter(compressedFile)
+	if err != nil {
+		return err
+	}
+	// copy the file to the zstd writer
+	_, err = io.Copy(zw, file)
+	if err != nil {
+		return err
+	}
+	// flush the zstd writer
+	if err := zw.Close(); err != nil {
+		return err
+	}
+	// remove the original file
+	if err := os.Remove(filepath); err != nil {
+		return err
+	}
+	return nil
+}
+
+func findLogs(logDir string, logPattern *regexp.Regexp) ([]string, error) {
+	var logFiles []string
+	err := filepath.WalkDir(logDir, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if !d.IsDir() && logPattern.MatchString(d.Name()) {
+			logFiles = append(logFiles, path)
+		}
+		return nil
+	})
+	return logFiles, err
+}
+
+func archiveLogs(logDir string, maxBackup int) error {
+	logPattern := regexp.MustCompile(`^access-\d{14}\.log$`)
+	logFiles, err := findLogs(logDir, logPattern)
+	if err != nil {
+		return err
+	}
+	if len(logFiles) >= maxBackup {
+		sort.Strings(logFiles)
+		// create archive log file with current date
+		date := time.Now().Format("20060102")
+		archiveFileName := fmt.Sprintf("archived-access-%s.log", date)
+		archiveFilePath := filepath.Join(logDir, archiveFileName)
+		// err if access-<date>.log.zst exists, and do nothing
+		if _, err := os.Stat(archiveFilePath + ".zst"); !os.IsNotExist(err) {
+			return fmt.Errorf("file %s exists", archiveFileName+".zst")
+		}
+		archiveFile, err := os.Create(archiveFilePath)
+		if err != nil {
+			return err
+		}
+		// concatenate `maxBackup` access-<datetime>.log files to access-<date>.log
+		for _, logFile := range logFiles {
+			src, err := os.Open(logFile)
+			if err != nil {
+				return err
+			}
+			_, err = io.Copy(archiveFile, src)
+			if err != nil {
+				return err
+			}
+			src.Close()
+		}
+		// compress access-<date>.log
+		if err := compressFile(archiveFilePath); err != nil {
+			return err
```

Comment:
`archiveFile` may remain open at this point. Better use `defer archiveFile.Close()` right after successfully opening.